### DISCUSSION TITLE
Gatsby v4 upgrade of gatsby-transformer-cloudinary Steps 1-5

### DISCUSSION THREAD

**Discussion by olavea** [olavea/gatsby-transformer-cloudinary-testing-testing-1-2-1-2#2] [Show and tell]
[Gatsby-v4 plugin upgrade Docs](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/#2-data-mutations-need-to-happen-during-sourcenodes-or-oncreatenode)

### The New Gatsby-v4 Way
`createTypes` is used inside `createSchemaCustomization`

Gatsby-v4 says: 👍 

[Look at `type CloudinaryAssetFixed` in /gatsby-node.js ](https://github.com/olavea/gatsby-transformer-cloudinary-testing-testing-1-2-1-2/blob/72677a509ae78302804c8c28bb8af025919e8e42/packages/gatsby-transformer-cloudinary/gatsby-node.js#L58)

### The Old Way
Instead of `createTypes`is used inside `sourceNodes`, which is The Old Way. But not used here 👍 

### The New Gatsby-v4 Way
The Tobbie-🐙 -Type is created in `createSchemaCustomization`and then referenced inside `onCreateNode`. Like this: `createAssetNodeFromFile(` is called inside `onCreateNode`

Gatsby-v4 says: 👍 

[Look at `createAssetNodeFromFile` in /gatsby-node.js ](https://github.com/olavea/gatsby-transformer-cloudinary-testing-testing-1-2-1-2/blob/72677a509ae78302804c8c28bb8af025919e8e42/packages/gatsby-transformer-cloudinary/gatsby-node.js#L238)

### The Old Way
Instead of a Tobbie 🐙  Node being created at the same time a resolver is created, which is The Old Way. But not used here 👍 

### The Old Way
`onPreInit` is setting 🌐  global state so that other functions can access that global state. Which is The Old Way. And used here  😉  

Gatsby-v4 says: 👎 

[Look at `onPreInit` in /gatsby-node.js ](https://github.com/olavea/gatsby-transformer-cloudinary-testing-testing-1-2-1-2/blob/72677a509ae78302804c8c28bb8af025919e8e42/packages/gatsby-transformer-cloudinary/gatsby-node.js#L248)

### The New Gatsby-v4 Way

```shell
yarn add gatsby-plugin-utils
```

The code will look something like this:

```js
// gatsby-plugin-root 🎄  / packages / gatsby-transformer-cloudinary / gatsby-node.js

let coreSupportsOnPluginInit = "unstable" | "stable" | undefined

try {
  const { isGatsbyNodeLifeCycleSupported } = require(`gatsby-plugin-utils`)
   if (isGatsbyNodeLifeCycleSupported(`onPluginInit`)) {
    coreSupportsOnPluginInit = "stable"
   } else if (isGatsbyNodeLifeCycleSupported(`unstable_onPluginInit`)) {
    coreSupportsOnPluginInit = "unstable" 
   }
} catch (error) {
  console.error(`Could not check if Gatsby supports onPluginInit lifecycle 🚴‍♀️  `)
}

let globalPluginOptions = {}

const initializeGlobalState = (_, pluginOptions) => {
  globalPluginOptions = pluginOptions
}

  if (coreSupportsOnPluginInit === "stable") {
    exports.onPluginInit = initializeGlobalState
  } else if (coreSupportsOnPluginInit === "unstable")  {
    exports.unstable_onPluginInit = initializeGlobalState 
  } else {
    exports.onPreBootstrap = initializeGlobalState
  }

// function in /packages/gatsby-transformer-cloudinary/options.js ?
``` 

[Look at `let coreSupportsOnPluginInit` in line 255 in /gatsby-node.js in my pull request ](https://github.com/olavea/gatsby-transformer-cloudinary-testing-testing-1-2-1-2/pull/8/files#)

[Gatsby-v4 plugin upgrade Docs](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/#2-data-mutations-need-to-happen-during-sourcenodes-or-oncreatenode)

**Comment by raae** [comment]
Great investigation so far! 
Interested in seeing results for point 4 and 5 as well.

**Comment by raae** [comment]
Seems to be som mutation happening that is no longer allowed as [described here](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/#dont-mutate-nodes-outside-of-expected-apis). 

Search for `__NODE` to find the code.

**Reply by raae** [reply]
Create an issue and fix it by changing the code as explainer in https://queen.raae.codes/emails/2022-05-26-direct-mutation/.
